os/net/mac/mac-sequence.c

```diff
@@ -45,79 +45,115 @@
 #include <string.h>
 
 #include "contiki-net.h"
+#include "lib/random.h"
 #include "net/mac/mac-sequence.h"
 #include "net/packetbuf.h"
 
-struct seqno {
-  linkaddr_t sender;
-  clock_time_t timestamp;
-  uint8_t seqno;
-};
+/* Log configuration */
+#include "sys/log.h"
+#define LOG_MODULE "mac-sequence"
+#define LOG_LEVEL LOG_LEVEL_MAC
 
 #ifdef NETSTACK_CONF_MAC_SEQNO_MAX_AGE
 #define SEQNO_MAX_AGE NETSTACK_CONF_MAC_SEQNO_MAX_AGE
 #else /* NETSTACK_CONF_MAC_SEQNO_MAX_AGE */
-#define SEQNO_MAX_AGE (20 * CLOCK_SECOND)
+#define SEQNO_MAX_AGE (20 /* seconds */)
 #endif /* NETSTACK_CONF_MAC_SEQNO_MAX_AGE */
 
-#ifdef NETSTACK_CONF_MAC_SEQNO_HISTORY
-#define MAX_SEQNOS NETSTACK_CONF_MAC_SEQNO_HISTORY
-#else /* NETSTACK_CONF_MAC_SEQNO_HISTORY */
-#define MAX_SEQNOS 16
-#endif /* NETSTACK_CONF_MAC_SEQNO_HISTORY */
-static struct seqno received_seqnos[MAX_SEQNOS];
+#ifdef MAC_SEQUENCE_CONF_WITH_BROADCAST_DEDUPLICATION
+#define WITH_BROADCAST_DEDUPLICATION MAC_SEQUENCE_CONF_WITH_BROADCAST_DEDUPLICATION
+#else /* MAC_SEQUENCE_CONF_WITH_BROADCAST_DEDUPLICATION */
+#define WITH_BROADCAST_DEDUPLICATION 0
+#endif /* MAC_SEQUENCE_CONF_WITH_BROADCAST_DEDUPLICATION */
+
+#define BEFORE(a, b) ((int16_t)((a) - (b)) < 0)
+
+struct deduplication_record {
+#if WITH_BROADCAST_DEDUPLICATION
+  uint16_t last_broadcast_seqno_timeout;
+#endif /* WITH_BROADCAST_DEDUPLICATION */
+  uint16_t last_unicast_seqno_timeout;
+#if WITH_BROADCAST_DEDUPLICATION
+  uint8_t last_broadcast_seqno;
+#endif /* WITH_BROADCAST_DEDUPLICATION */
+  uint8_t last_unicast_seqno;
+};
+
+static uint8_t mac_dsn;
+NBR_TABLE(struct deduplication_record, deduplication_table);
 
 /*---------------------------------------------------------------------------*/
-int
-mac_sequence_is_duplicate(void)
+void
+mac_sequence_init(void)
 {
-  int i;
-  clock_time_t now = clock_time();
-
-  /*
-   * Check for duplicate packet by comparing the sequence number of the incoming
-   * packet with the last few ones we saw.
-   */
-  for(i = 0; i < MAX_SEQNOS; ++i) {
-    if(linkaddr_cmp(packetbuf_addr(PACKETBUF_ADDR_SENDER),
-                    &received_seqnos[i].sender)) {
-      if(packetbuf_attr(PACKETBUF_ATTR_MAC_SEQNO) == received_seqnos[i].seqno) {
-#if SEQNO_MAX_AGE > 0
-        if(now - received_seqnos[i].timestamp <= SEQNO_MAX_AGE) {
-          /* Duplicate packet. */
-          return 1;
-        }
-#else /* SEQNO_MAX_AGE > 0 */
-        return 1;
-#endif /* SEQNO_MAX_AGE > 0 */
-      }
-      break;
-    }
-  }
-  return 0;
+  mac_dsn = random_rand();
+  nbr_table_register(deduplication_table, NULL);
 }
 /*---------------------------------------------------------------------------*/
 void
-mac_sequence_register_seqno(void)
+mac_sequence_set_dsn(void)
 {
-  int i, j;
+  packetbuf_set_attr(PACKETBUF_ATTR_MAC_SEQNO, mac_dsn++);
+}
+/*---------------------------------------------------------------------------*/
+enum mac_sequence_result
+mac_sequence_is_duplicate(void)
+{
+  uint16_t now;
```

Comment:
Everything is built as C99, put the variable declarations at their initialization.